src/Microsoft.AspNetCore.SignalR.Client/HubConnection.cs

```diff
@@ -204,69 +217,114 @@ private void Shutdown(Exception ex = null)
-
             // If the invocation hasn't been cancelled, dispatch the result
             if (!irq.CancellationToken.IsCancellationRequested)
             {
-                irq.Registration.Dispose();
+                irq.ReceiveResult(result.Item);
+            }
+        }
+
+        private void DispatchInvocationCompletion(CompletionMessage completion, InvocationRequest irq, CancellationToken cancellationToken)
+        {
+            _logger.LogInformation("Received Completion for Invocation #{0}", completion.InvocationId);
+
+            if (cancellationToken.IsCancellationRequested)
+            {
+                return;
+            }
+
+            // If the invocation hasn't been cancelled, dispatch the completion
+            if (!irq.CancellationToken.IsCancellationRequested)
+            {
+                if (!string.IsNullOrEmpty(completion.Error))
+                {
+                    irq.Complete(new Exception(completion.Error));
+                }
+                else
+                {
+                    if (completion.HasResult)
+                    {
+                        irq.ReceiveResult(completion.Result);
+                    }
 
-                // Complete the request based on the result
-                // TODO: the TrySetXYZ methods will cause continuations attached to the Task to run, so we should dispatch to a sync context or thread pool.
-                if (!string.IsNullOrEmpty(result.Error))
+                    irq.Complete();
+                }
+            }
+        }
+
+        private void RemoveRequest(string invocationId)
+        {
+            lock (_pendingCallsLock)
+            {
+                if (!_pendingCalls.ContainsKey(invocationId))
                 {
-                    _logger.LogInformation("Completing Invocation #{0} with error: {1}", result.Id, result.Error);
-                    irq.Completion.TrySetException(new Exception(result.Error));
+                    _logger.LogWarning("Duplicate request to remove invocation {invocationId} from the queue ignored", invocationId);
                 }
                 else
                 {
-                    _logger.LogInformation("Completing Invocation #{0} with result of type: {1}", result.Id, result.Result?.GetType()?.FullName ?? "<<void>>");
-                    irq.Completion.TrySetResult(result.Result);
+                    _pendingCalls.Remove(invocationId);
                 }
             }
         }
 
+        private void EnqueueRequest(string invocationId, InvocationRequest irq)
+        {
+            if (_connectionActive.IsCancellationRequested)
+            {
+                throw new InvalidOperationException("Connection has been terminated.");
+            }
+
+            lock (_pendingCallsLock)
+            {
+                if (_pendingCalls.ContainsKey(invocationId))
+                {
+                    throw new InvalidOperationException("An invocation with this ID has already been queued");
```

Comment:
I've changed back to a ConcurrentDictionary, which means `TryAdd` gives this to us for free. I'd rather manifest errors like this when they happen, whereas `Debug.Assert` is hidden from the user in Release builds.

It's the difference between getting a bug report that says "After sending 4 billion messages, I got an Exception saying duplicate invocation IDs were generated" and getting one that says "After sending 4 billion messages, weird stuff started happening" :). Yeah, it's super unlikely, but I'd rather know what's happening.